File: 2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.java
class Solution {
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        int[] degree = new int[n];
        List<List<Integer>> graph = new ArrayList<>();
        List<Set<Integer>> answer = new ArrayList<>();
        List<List<Integer>> result = new ArrayList<>();
        Queue<Integer> que = new ArrayDeque<>();

        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
            answer.add(new HashSet<>());
        }

        for (int i = 0; i < edges.length; i++) {
            degree[edges[i][1]]++;
            graph.get(edges[i][0]).add(edges[i][1]);
        }

        for (int i = 0; i < n; i++) {
            if (degree[i] == 0) que.add(i);
        }

        while (!que.isEmpty()) {
            int now = que.poll();
            for (int i = 0; i < graph.get(now).size(); i++) {
                degree[graph.get(now).get(i)]--;
                answer.get(graph.get(now).get(i)).addAll(answer.get(now));
                answer.get(graph.get(now).get(i)).add(now);
                if (degree[graph.get(now).get(i)] == 0) {
                    que.add(graph.get(now).get(i));
                }
            }
        }

        for (int i = 0; i < n; i++) {
            result.add(answer.get(i).stream().sorted().collect(Collectors.toList()));
        }

        return result;
    }
}

File: 2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.py
class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        degree = [0 for i in range(n)]
        graph = [[] for i in range(n)]
        answer = [set() for i in range(n)]
        result = list()
        que = deque()

        for u, v in edges:
            degree[v] += 1
            graph[u].append(v)

        for i in range(n):
            if degree[i] == 0:
                que.append(i)

        while que:
            now = que.popleft()
            for next_point in graph[now]:
                degree[next_point] -= 1
                answer[next_point].update(answer[now])
                answer[next_point].add(now)
                if degree[next_point] == 0:
                    que.append(next_point)

        for i in range(n):
            result.append(sorted(answer[i]))

        return result